### PR TITLE
Run the test suite

### PR DESCRIPTION
**Pull request by jakirkham**
We had tried running the test suite before, but had some kind of issue. Suspect it was running out of memory, but don't have a log available. However, this was nearly a year ago and many things have changed. So am giving it another try to see if it can run.

### CODE REVIEW

**Comment by jakirkham** [comment]
Well, it seems the tests on Mac with OpenMP are doomed to fail at this point. Others seem to have run into the same issue with single threaded OpenBLAS ( https://github.com/xianyi/OpenBLAS/issues/1141 ), which is close enough to our use case here. So it sounds like a memory issue with OpenBLAS triggered by SPAMS usage.

I think if we rebuilt OpenBLAS to use a reasonably large max number of threads we can fix the underlying memory issues as noted in the [FAQ]( https://github.com/xianyi/OpenBLAS/wiki/Faq#usage-1 ) and the aforementioned issue. However lack the time at present to explore this further. Short term am disabling the test suite on Mac. Running the test suite on Linux is already an improvement over not running them at all.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -47,6 +50,19 @@ requirements:
...
 test:
+  source_files:
+    # Test data
+    - boat.png
+    - lena.png           # [linux]
+    # Test suite
+    - test_decomp.py     # [linux]
+    - test_dictLearn.py
+    - test_linalg.py
+    - test_prox.py       # [linux]
```
Comment (jakirkham) [inline recipe/meta.yaml]: As OpenBLAS on macOS seems to have some sensitive memory issues (segfaults), we skip some of these tests. This should be resolvable with a rebuild of OpenBLAS, but will require more investigation. Fortunately with the test suite simply excluding the test module or test data is sufficient to skip that port. So we skip copying over any of the failing portions on macOS.

**Comment by jakirkham** [comment]
Going with this for now as it vastly improves testing of the package. Can iterate on handling the tests skipped on macOS once we have tried rebuilding OpenBLAS.

**Comment by zym1010** [comment]
@jakirkham  I ran the test suite, and it ran flawlessly on Linux.

However, I should point out that, the original test suite simply won't fail at all. There's no `assert` in it....

Anyway, knowing that it would run successfully is a good thing.

As for OpenMP + OpenBLAS, is that possible to compile OpenMP enabled OpenBLAS on Linux? Or is it the case that we must make a hard choice between using and not using OpenMP? Really hate those warnings (`Detect OpenMP Loop and this application may hang. Please rebuild the library with USE_OPENMP=1 option.`) when using multithread...

**Comment by jakirkham** [comment]
> However, I should point out that, the original test suite simply won't fail at all. There's no assert in it....

Really? Not so sure. See all the little red X's before we got this passing. 😉

> Anyway, knowing that it would run successfully is a good thing.

Indeed.

> As for OpenMP + OpenBLAS, is that possible to compile OpenMP enabled OpenBLAS on Linux?

I don't think that is a good idea.

GNU's OpenMP implementation is [not fork safe]( http://bisqwit.iki.fi/story/howto/openmp/#OpenmpAndFork ). Intel's is for comparison. Not sure about the situation with Clang. So this by no means part of the OpenMP spec. Given Python's `multiprocessing` module uses forking extensively, lots of stuff will start breaking unexpectedly. There are potential workarounds (particularly in newer Pythons), but they don't exist in Python 2. Here's a [related bug report]( https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58378 ). Have experienced this personally before and it is a nightmare to debug. In the end, I found it just easier to disable OpenMP on OpenBLAS and move on with life.

That all said, I do sympathize with you that this is very annoying.

> Or is it the case that we must make a hard choice between using and not using OpenMP? Really hate those warnings (`Detect OpenMP Loop and this application may hang. Please rebuild the library with USE_OPENMP=1 option.`) when using multithread...

Here I just [chose]( https://github.com/conda-forge/python-spams-feedstock/blob/dd6e6e777c746abbf5ea695a853900c792885283/recipe/run_test.py#L6-L7 ) not to use OpenBLAS parallelism when using OpenMP parallelism with SPAMS. I think one could go about this the other way instead if they wished. TBH I don't think this is an unreasonable choice to make to avoid contention related issues.

**Comment by zym1010** [comment]
@jakirkham  thanks for the explanation on OpenBLAS issue! However, I just wonder if disabling multithreading, how much performance will be lost... Have you made any comparison? TBH, I exclusively use spams for inference, and it seems to me, I don't see any such warnings.

**Comment by jakirkham** [comment]
I did a comparison a couple of years back, but it was pretty specific to our use case. That said, I can try to search for it if it is of interest.

**Comment by zym1010** [comment]
@jakirkham  never mind. seems that disabling multithreading doesn't give me much performance loss, and without multithreading, it's still possible to use more than one core.

**Comment by jakirkham** [comment]
> That's why I hate those OpenMP programs, such as Torch...

😆 Me too.

**Comment by jakirkham** [comment]
FWIW we also sandboxed SPAMS early on. That way we could use different settings at different points. Maybe a similar strategy will work for you.

> seems that disabling multithreading doesn't give me much performance loss, and without multithreading, it's still possible to use more than one core.

That's what I remember. Parallelism in SPAMS is mainly working with small batches. So parallelism on this level is going to give you the biggest boost.

**Comment by zym1010** [comment]
so the reason we can use more than one core is that calls to openblas implicitly uses multiple cores, right?

**Comment by jakirkham** [comment]
OpenBLAS does use some autovectorization instructions if that is what you are meaning.

**Pull request merged**